File: built-in plugins dsp/mb-mix-parallel.cpp
/**
 * MB Parallel Processor
 * Category : effect
 * Type     : mixing
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Parallel processing chain with dry/wet blend and sidechain
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MIX_PARALLEL_H
#define MB_MIX_PARALLEL_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMixParallel : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-mix-parallel";
    static constexpr const char* PLUGIN_NAME    = "MB Parallel Processor";
    static constexpr const char* PLUGIN_TYPE    = "mixing";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float blend = 0.5f;  // range [0, 1]
    float input = 0f;  // range [-12, 12]
    float output = 0.8f;  // range [0, 1]
    };

    MbMixParallel() = default;
    ~MbMixParallel() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.blend = std::clamp(params.blend, 0f, 1f);
        params.input = std::clamp(params.input, -12f, 12f);
        params.output = std::clamp(params.output, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Parallel Processor
        return input;
    }
};

#endif // MB_MIX_PARALLEL_H

File: built-in plugins dsp/mb-mix-parallel.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbMixParallelPlugin: PluginDefinition = { id: 'mb-mix-parallel', slug: 'mb-mix-parallel', name: 'MB Parallel Processor', category: 'effect', type: 'mixing' as any, version: '1.0.0', description: 'Parallel processing chain with dry/wet blend and sidechain', author: 'Max Booster', grade: 'A', parameters: [{ id: 'blend', name: 'Blend', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'input', name: 'Input Gain', type: 'float', defaultValue: 0, minValue: -12, maxValue: 12, automatable: true }, { id: 'output', name: 'Output', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { blend: 0.5, input: 0, output: 0.8 } };

export default MbMixParallelPlugin;
